src/python/pants/build_graph/target.py

```diff
@@ -471,9 +471,12 @@ def transitive_invalidation_hash(self, fingerprint_strategy=None, depth=0):
       hasher = sha1()
 
       def dep_hash_iter():
-        for dep in self.dependencies:
+        for dep in fingerprint_strategy.dependencies(self):
           try:
-            dep_hash = dep.transitive_invalidation_hash(fingerprint_strategy, depth=depth+1)
+            if fingerprint_strategy.direct(self):
```

Comment:
The recursion here makes this tricky. I'd like to see a few test cases using targets where strict_deps=False whose dependencies have strict_deps=True.

Consider
```
A => B
B => C
C => D
where B has strict_deps=True
```

Currently, the dependencies used in A's compile classpath will be B, C, D. A's transitive fingerprint will also be based on those three targets.

With this change as it is, the classpath will remain the same, but the fingerprint will not include D because B is strict_deps=True and direct(B) will return True.

We should resolve this discrepancy. I'd also like to see a test case that looks like that and specifies the expected behavior.

We can resolve this by either having direct only apply to the top level target, ie with depth 0, or by changing the compile classpath. I think having direct only apply to the top level target would be better.